beetsplug/mbsubmit.py

```diff
@@ -18,31 +18,334 @@
+            # MusicBrainz redirects to this endpoint after successfully adding a
+            #  release. The ID of the new release is provided, completing the flow.
+            self._complete_add(args)
+        else:
+            self._response(404, "Unknown action.")
+
+    def _get_task_from_args(self, args) -> Optional[CreateReleaseTask]:
+        # Try to get the token from query args, try to decode it, and try to find the
+        #  associated CreateReleaseTask.
+        if "token" in args:
+            token = args["token"][0]
+            try:
+                payload = jwt.decode(
+                    token,
+                    self._plugin.jwt_key,
+                    algorithms=self._plugin.jwt_algorithm,
+                )
+                if (
+                    "task_key" in payload
+                    and payload["task_key"] in self._plugin.create_release_tasks
+                ):
+                    return self._plugin.create_release_tasks[
+                        payload["task_key"]
+                    ]
+                else:
+                    self._response(404, "task_key not found.")
+            except jwt.InvalidTokenError:
+                self._response(400, "Invalid token.")
+        else:
+            self._response(400, "Token missing.")
+        return None
+
+    def _add(self, args):
+        if (task := self._get_task_from_args(args)) is None:
+            return
+
+        template = _form_template.format(
+            action=escape("https://musicbrainz.org/release/add"),
+            form_data="".join(
+                [
+                    _form_input_template.format(
+                        name=escape(str(k)), value=escape(str(v))
+                    )
+                    for k, v in task.formdata.items()
+                ]
+            ),
+        )
+        task.browser_opened = True
+        self._response(200, template, content_type="text/html")
+
+    def _complete_add(self, args):
+        if (task := self._get_task_from_args(args)) is None:
+            return
+
+        if "release_mbid" not in args:
+            self._response(400, "release_mbid missing.")
+            return
+
+        release_mbid = args["release_mbid"][0]
+
+        task.result_release_mbid = release_mbid
+        self._response(
+            200,
+            f"Release {release_mbid} added. You can close this browser window now and "
+            "return to beets.",
+        )
+
+    def _response(
+        self, code: int, content: str = "", content_type: str = "text/plain"
+    ):
+        self.send_response(code)
+        self.send_header("Content-Type", content_type)
+        self.send_header("Cache-Control", "max-age=0")
+        self.end_headers()
+        self.wfile.write(content.encode())
+
+
+def join_phrase(i, total):
+    if i < total - 2:
+        return ", "
+    elif i < total - 1:
+        return " & "
+    else:
+        return ""
+
+
+def build_formdata(items: list, redirect_uri: Optional[str]):
```

Comment:
Your use of typings is inconsistent. Would you consider typing all of these functions? Both the parameters and return types?